### PR TITLE
[ECO-5116] Switch to V2 Chat API

### PR DESCRIPTION
**Pull request by maratal**
I've temporarily switched to the branch of ably-cocoa [PR](https://github.com/ably/ably-cocoa/actions/runs/11954211970?pr=1998) It should be merged before this PR so I could remove tmp commit.

## Summary by CodeRabbit

- **New Features**
	- Enhanced message handling with new properties for reactions and message identifiers.
	- Introduced support for the new chat API version.
	- Updated event handling to reflect new message events.

- **Bug Fixes**
	- Corrected message identifier usage in various components.

- **Documentation**
	- Updated tests to align with the new message structure and properties.

- **Chores**
	- Updated dependencies to point to a new feature branch for improved chat support.

### CODE REVIEW

**Comment by umair-ably** [comment]
Looks good. Only things to add would be:
- We need to remove CHA-M5j (I've confirmed this with Andy, we just need to remove the few lines of code here - search for spec point in code)
- We can get rid of the timeserial comparison functions e.g. isBefore, isAfter, isEqual... as we can now just compare them lexicographically e.g. using `>` or `==`

**Comment by maratal** [comment]
[This](https://github.com/ably-labs/ably-chat-swift/pull/122/commits/5fa1257dea0ce0b194eb4e8f3b9f158c2755780a)? @umair-ably

**Comment by umair-ably** [comment]
> [This](https://github.com/ably-labs/ably-chat-swift/pull/122/commits/5fa1257dea0ce0b194eb4e8f3b9f158c2755780a)? @umair-ably

yeah that's right. I'd also get rid of the actual functions itself e.g. `public func isBefore(_ otherMessage: Message) throws -> Bool`... we can directly use the operators instead of nesting them inside these functions

**Comment by maratal** [comment]
@umair-ably there are also changes in the realtime message structure, should I update ably-cocoa as well then?

**Comment by maratal** [comment]
@umair-ably I've pushed draft version of all requested changes (apart from those in cocoa). You can preview it now.

File: Sources/AblyChat/ChatAPI.swift [hunk 1]
```diff
@@ -3,19 +3,20 @@ import Ably
 internal final class ChatAPI: Sendable {
     private let realtime: RealtimeClient
     private let apiVersion = "/chat/v1"
+    private let apiVersionV2 = "/chat/v2" // TODO: remove v1 after full transition to v2
```
Comment (umair-ably) [inline Sources/AblyChat/ChatAPI.swift:6]: I think we're okay to do this as part of this PR

File: AblyChat.xcworkspace/xcshareddata/swiftpm/Package.resolved [hunk 2]
```diff
@@ -1,13 +1,13 @@
 {
-  "originHash" : "1ad2d7338668d15feccbf564582941161acd47349bfca8f34374e11c69677ae8",
+  "originHash" : "37204b528044b0c8af30d6383e4706728b14b8da503c5f0b346e4cbd6533d919",
```
[the remark this answers is not in the document]
Comment (maratal) [reply AblyChat.xcworkspace/xcshareddata/swiftpm/Package.resolved:2]: Ignore changes in all Package.resolved files

**Comment by maratal** [comment]
@coderabbitai pause

File: Sources/AblyChat/ChatAPI.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (maratal) [reply Sources/AblyChat/ChatAPI.swift:6]: This is by design for this PR.

File: Sources/AblyChat/Events.swift [hunk 3]
```diff
@@ -1,5 +1,28 @@
...
+import Ably
+
+public enum MessageEvent: String, Codable, Sendable {
+    case create = "message.create"
+
+    internal static func fromAction(_ action: ARTMessageAction) -> Self? {
+        switch action {
+        case .create:
+            .create
+        // ignore any other actions except `chat.message` for now
```
[the remark this answers is not in the document]
Comment (maratal) [reply Sources/AblyChat/Events.swift]: It's allowed in swift 6.

File: Sources/AblyChat/ChatAPI.swift [hunk 4]
```diff
@@ -3,19 +3,20 @@ import Ably
...
     // (CHA-M6) Messages should be queryable from a paginated REST API.
     internal func getMessages(roomId: String, params: QueryOptions) async throws -> any PaginatedResult<Message> {
-        let endpoint = "\(apiVersion)/rooms/\(roomId)/messages"
+        let endpoint = "\(apiVersionV2)/rooms/\(roomId)/messages"
         return try await makePaginatedRequest(endpoint, params: params.asQueryItems())
     }
 
     internal struct SendMessageResponse: Codable {
-        internal let timeserial: String
+        internal let serial: String
```
[the remark this answers is not in the document]
Comment (maratal) [reply Sources/AblyChat/ChatAPI.swift:19]: fixed

File: Example/AblyChatExample/Mocks/Misc.swift [hunk 5]
```diff
@@ -11,7 +11,8 @@ final class MockMessagesPaginatedResult: PaginatedResult {
     var items: [T] {
         Array(repeating: 0, count: numberOfMockMessages).map { _ in
             Message(
-                timeserial: "\(Date().timeIntervalSince1970)",
+                serial: "\(Date().timeIntervalSince1970)",
```
[the remark this answers is not in the document]
Comment (maratal) [reply Example/AblyChatExample/Mocks/Misc.swift:14]: It's ok for tests here.

File: Sources/AblyChat/Message.swift [hunk 6]
```diff
@@ -6,18 +6,20 @@ public typealias MessageMetadata = Metadata
 // (CHA-M2) A Message corresponds to a single message in a chat room. This is analogous to a single user-specified message on an Ably channel (NOTE: not a ProtocolMessage).
 public struct Message: Sendable, Codable, Identifiable, Equatable {
     // id to meet Identifiable conformance. 2 messages in the same channel cannot have the same timeserial.
-    public var id: String { timeserial }
+    public var id: String { serial }
 
-    public var timeserial: String
+    public var serial: String
+    public var latestAction: MessageEvent
```
[the remark this answers is not in the document]
Comment (maratal) [reply Sources/AblyChat/Message.swift:12]: This is ok for now.

File: Sources/AblyChat/ChatAPI.swift (hunk 1 above)
Comment (maratal) [reply Sources/AblyChat/ChatAPI.swift:6]: You mean replace to v2 everywhere?

Comment (umair-ably) [reply Sources/AblyChat/ChatAPI.swift:6]: as in I don't see why we need both apiVersion and apiVersionV2? what do you mean by full transition to V2?

File: Sources/AblyChat/Events.swift [hunk 7]
```diff
@@ -1,5 +1,28 @@
-internal enum MessageEvent: String {
-    case created = "message.created"
+import Ably
+
+public enum MessageEvent: String, Codable, Sendable {
```
Comment (umair-ably) [inline Sources/AblyChat/Events.swift:3]: I think we should rename this to `MessageAction`

File: Sources/AblyChat/Events.swift [hunk 8]
```diff
@@ -1,5 +1,28 @@
...
+             .annotationDelete,
+             .metaOccupancy:
+            nil
+        @unknown default:
+            nil
+        }
+    }
+}
+
+internal enum MessageFilter: String, Sendable {
```
Comment (umair-ably) [inline Sources/AblyChat/Events.swift:24]: and this should remain as `MessageEvent`

**Comment by umair-ably** [comment]
actually ignore that... this wasn't in the doc as it's part of edit/delete but from looking at ably-js we're going to have a new type which also sounds confusingly similar... we don't need it for this PR but the solution will grow into something like this:

Screenshot 2024-11-22 at 00 36 00

We should just follow the same naming conventions now to avoid both confusion and the headache of coming up with our own names that make sense

File: Sources/AblyChat/ChatAPI.swift (hunk 1 above)
Comment (maratal) [reply Sources/AblyChat/ChatAPI.swift:6]: "Move to V2 endpoints for history and sending messages." in the doc.

Comment (maratal) [reply Sources/AblyChat/ChatAPI.swift:6]: It's 404 for occupancy
Screenshot 2024-11-22 at 01 59 39

Comment (umair-ably) [reply Sources/AblyChat/ChatAPI.swift:6]: ahh cool, makes sense!

Comment (maratal) [reply Sources/AblyChat/ChatAPI.swift:6]: So is this that one thing to approval, or renaming?

File: Sources/AblyChat/Events.swift (hunk 8 above)
Comment (umair-ably) [reply Sources/AblyChat/Events.swift:24]: replaced by larger comment with screenshot

File: Sources/AblyChat/Events.swift (hunk 7 above)
Comment (umair-ably) [reply Sources/AblyChat/Events.swift:3]: replaced by larger comment with screenshot

**Review by umair-ably — APPROVED** [review]
CI failing is due to https://ably-real-time.slack.com/archives/C03JDBVM5MY/p1732287002273979 and not this PR... approving but wait until CI passes again... Should CI fail again from next week due to expected timeserial changes, then contact @AndyTWF for how to resolve them in line with expected behaviour

**Pull request merged**